src/Std/Data/HashMap/Lemmas.lean

```diff
@@ -1269,4 +1269,330 @@ theorem isEmpty_unitOfList [EquivBEq α] [LawfulHashable α]
 
 end
 
+section Alter
+
+variable {m : HashMap α β}
+
+@[simp]
+theorem isEmpty_alter [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β} :
+    (alter m k f).isEmpty ↔ (m.erase k).isEmpty ∧ f (get? m k) = none :=
+  DHashMap.Const.isEmpty_alter
+
+theorem contains_alter [EquivBEq α] [LawfulHashable α] {k k': α} {f : Option β → Option β} :
+    (alter m k f).contains k' =
+    if k == k' then (f (get? m k)).isSome else m.contains k' :=
+  DHashMap.Const.contains_alter
+
+theorem mem_alter [EquivBEq α] [LawfulHashable α] {k k': α} {f : Option β → Option β} :
+    k' ∈ alter m k f ↔ if k == k' then (f (get? m k)).isSome = true else k' ∈ m :=
+  DHashMap.Const.mem_alter
+
+theorem mem_alter_of_beq [EquivBEq α] [LawfulHashable α] {k k': α} {f : Option β → Option β}
+    (h : k == k') : k' ∈ alter m k f ↔ (f (get? m k)).isSome :=
+  DHashMap.Const.mem_alter_of_beq h
+
+@[simp]
+theorem contains_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β} :
+    (alter m k f).contains k = (f (get? m k)).isSome :=
+  DHashMap.Const.contains_alter_self
+
+@[simp]
+theorem mem_alter_self [EquivBEq α] [LawfulHashable α] {k : α} {f : Option β → Option β} :
+    k ∈ alter m k f ↔ (f (get? m k)).isSome :=
+  DHashMap.Const.mem_alter_self
+
+theorem contains_alter_of_beq_eq_false [EquivBEq α] [LawfulHashable α] {k k' : α}
+    {f : Option β → Option β} (h : (k == k') = false) :
+    (alter m k f).contains k' = m.contains k' :=
+  DHashMap.Const.contains_alter_of_beq_eq_false h
+
+theorem mem_alter_of_beq_eq_false [EquivBEq α] [LawfulHashable α] {k k' : α}
+    {f : Option β → Option β} (h : (k == k') = false) : k' ∈ alter m k f ↔ k' ∈ m :=
+  DHashMap.Const.mem_alter_of_beq_eq_false h
+
+theorem size_alter [LawfulBEq α] {k : α} {f : Option β → Option β} :
+    (alter m k f).size =
+      if m.contains k && (f (get? m k)).isNone then
+        m.size - 1
+      else if !m.contains k && (f (get? m k)).isSome then
+        m.size + 1
+      else
+        m.size :=
+  DHashMap.Const.size_alter
+
+theorem size_alter_eq_add_one [LawfulBEq α] {k : α} {f : Option β → Option β}
+    (h : m.contains k = false) (h': (f (get? m k)).isSome) :
+    (alter m k f).size = m.size + 1 :=
+  DHashMap.Const.size_alter_eq_add_one h h'
+
+theorem size_alter_eq_sub_one [LawfulBEq α] {k : α} {f : Option β → Option β}
+    (h : m.contains k) (h': (f (get? m k)).isNone) :
+    (alter m k f).size = m.size - 1 :=
+  DHashMap.Const.size_alter_eq_sub_one h h'
+
+theorem size_alter_eq_self [LawfulBEq α] {k : α} {f : Option β → Option β}
+    (h : m.contains k = false) (h': (f (get? m k)).isNone) :
+    (alter m k f).size = m.size :=
+  DHashMap.Const.size_alter_eq_self h h'
+
+theorem size_alter_eq_self' [LawfulBEq α] {k : α} {f : Option β → Option β}
```

Comment:
We're trying not to intoduce new `'` into theorem names where possible. Please disambiguate using the first hypothesis that is different.

Reply:
What do you think of `size_alter_eq_self_of_contains` and `size_alter_eq_self_of_contains_eq_false`?

An alternative would be to just have one lemma that covers both statements:

```lemma4
theorem size_alter_eq_self [LawfulBEq α] {k : α} {f : Option β → Option β}
    (h : m.contains k = f (get? m k)).isSome) :
    (alter m k f).size = m.size := sorry
```